### ISSUE TITLE
broken with tensorflow 1.6

### ISSUE THREAD

**Issue by hughsalimbeni** [GPflow/GPflow#677]
tensorflow 1.6 does not currently work (fails with Adam), but the docs say 1.4+ is supported. This can cause problems for users (e.g. https://github.com/ICL-SML/Doubly-Stochastic-DGP/issues/10).

**Comment by mrksr** [comment]
The issue has been discussed in the [tensorflow repo](https://github.com/tensorflow/tensorflow/issues/8057#issuecomment-284000067). Without digging into the details, I tried the workaround proposed there which seems to do the trick. I would not call it a solution however.

```patch
diff --git a/gpflow/training/tensorflow_optimizer.py b/gpflow/training/tensorflow_optimizer.py
index 2ecef24..9fd1af6 100644
--- a/gpflow/training/tensorflow_optimizer.py
+++ b/gpflow/training/tensorflow_optimizer.py
@@ -87,6 +87,8 @@ class _TensorFlowOptimizer(optimizer.Optimizer):
                         yield slot
         extra_vars = [v for v in self.optimizer.__dict__.values() if isinstance(v, tf.Variable)]
         optimizer_vars = list(get_optimizer_slots())
+        if isinstance(self.optimizer, tf.train.AdamOptimizer):
+            optimizer_vars.extend(self.optimizer._get_beta_accumulators())
         full_var_list = list(set(optimizer_vars + extra_vars))
         misc.initialize_variables(full_var_list, session=session, force=False)
```

**Comment by namoshizun** [comment]
I ran into the same problem when just now tried to run the example in this [tutorial](http://gpflow.readthedocs.io/en/latest/notebooks/mcmc.html). The workaround solution @mrksr gave works for me.

**Comment by rhaps0dy** [comment]
What about using `self.optimizer.variables()`, instead of which was [introduced in Tensorflow 1.5](https://www.tensorflow.org/versions/r1.5/api_docs/python/tf/train/AdamOptimizer)? Instead of `get_optimizer_slots`.

**Comment by hughsalimbeni** [comment]
@rhaps0dy that seems like a good way to go. I think this should go in asap as it's a bit frustrating/off-putting for new users if it doesn't work with a fresh tensorflow. Any objections @awav?

**Comment by imsrgadich** [comment]
@rhaps0dy's code still gave me the error. but @mrksr code worked!

**Comment by awav** [comment]
Solved #714
